paddle/fluid/framework/details/multi_devices_graph_pass.cc

```diff
@@ -271,7 +271,15 @@ bool MultiDevSSAGraphBuilderBase::UseGPU() const {
   return use_gpu;
 }
 
-bool MultiDevSSAGraphBuilderBase::NeedCollectiveOps() const {
+bool MultiDevSSAGraphBuilderBase::NeedCollectiveOps(
+    std::vector<ir::Node *> ops) const {
+  for (auto *node : ops) {
+    if (node->Op()->Type() == "allreduce") {
+      // HACK: if graph have allreduce op, do not append automatic
+      // allreduce allreduce_op_handles
```

Comment:
what's the long term plan?

Reply:
> what's the reason for doing this?

To enable users to create arbitrary multi-device/distributed models at will, See the example above, currently the loss averaging is done by internal passes, if user want to average gradients other than loss, they may add it there, or people can run allreduce to any variable to achieve functions like "syncbn".

> what's the long term plan?

`multi_devices_graph_pass` only do replicate graph to multiple devices, python side `compiler.py:with_data_parallel` will add allreduce op for the program.